### PR TITLE
[Merged by Bors] - Feat (Analysis/Normed/Ring/Seminorm): add lemmas on equivalence of MulRingNorms

### PR DESCRIPTION
**Pull request by NirvanaC93**
Include the following lemmas on equivalence of `MulRingNorms`. 
```lean
lemma f_of_abs_eq_f (x : ℤ) : f (Int.natAbs x) = f x
lemma NormRat_eq_on_Int_iff_eq_on_Nat : (∀ n : ℕ , f n = g n) ↔ (∀ n : ℤ , f n = g n)
lemma NormRat_eq_iff_eq_on_Nat : (∀ n : ℕ , f n = g n) ↔ f = g
lemma NormRat_equiv_iff_equiv_on_Nat : (∃ c : ℝ, 0 < c ∧ (∀ n : ℕ , (f n)^c = g n)) ↔ f.equiv g
```

Co-authored-by:

David Kurniadi Angdinata [dka31@cantab.ac.uk](mailto:dka31@cantab.ac.uk)
Fabrizio Barroero [fabrizio.barroero@uniroma3.it](mailto:fabrizio.barroero@uniroma3.it)
Laura Capuano [laura.capuano@uniroma3.it](mailto:laura.capuano@uniroma3.it)
Nirvana Coppola [nirvanac93@gmail.com](mailto:nirvanac93@gmail.com)
María Inés de Frutos Fernández [maria.defrutos@uam.es](mailto:maria.defrutos@uam.es)
Silvain Rideau-Kikuchi [silvain.rideau-kikuchi@ens.fr](mailto:silvain.rideau-kikuchi@ens.fr)
Sam van Gool [vangool@irif.fr](mailto:vangool@irif.fr)
Francesco Veneziano [veneziano@dima.unige.it](mailto:veneziano@dima.unige.it)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (AntoineChambert-Loir) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:424]: Once on has Ostrowski, one knows that if there is n such that f n \neq 1 and g n \neq 1, then f and g are equivalent.

**Comment by AntoineChambert-Loir** [comment]
I read the statements. They are useful within a proof of Ostrowski but I am not sure they are really interesting besides that. I wonder whether they could be included in a Ostrowski file, as private lemmas.

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:424]: Hi! Yes, and we use this Lemma in Ostrowski's proof. :)

**Comment by mariainesdff** [comment]
> I read the statements. They are useful within a proof of Ostrowski but I am not sure they are really interesting besides that. I wonder whether they could be included in a Ostrowski file, as private lemmas.

I agree with moving them to a file specific to the rationals (but I am not sure about making them private lemmas).

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:393]: Where is this variable used?

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:397]: This should be renamed. Maybe to `MulRingNorm.apply_natAbs_eq`?

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:398]: You probably want to open the `Int` namespace.

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:404]: I am not sure about the name of this and the next two lemmas.

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:397]: Also, this should work for `f : MulRingNorm R` over any ring `R`.

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:420]:
```suggestion
  rw [← Rat.num_div_den z, map_div₀, map_div₀, h, NormRat_eq_on_Int_iff_eq_on_Nat.mp h]
```

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:428]:
```suggestion
  · intro ⟨c, hc, h⟩
```

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:435]:
```suggestion
  · intro ⟨c, hc, h⟩
```

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:439]:
```suggestion
  exact ⟨c, ⟨hc, fun n ↦ by rw [← h]⟩⟩
```
You may even golf the lemma proof a bit more, if you like, replacing the `constructor` on line 426 by
```
refine ⟨?_, fun ⟨c, hc, h⟩ ↦ ⟨c, ⟨hc, fun n ↦ by rw [← h]⟩⟩⟩
```

**Review by mariainesdff — CHANGES_REQUESTED** [review]

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:393]: You're right, we never use it here. But seeing the next suggestion we can use it :D

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:397]: I like your suggestion. As for the generalization to any R, then that'd justify the variable R :D

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:398]: Done

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:404]: Any suggestion is welcome!

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:439]: Changes implemented!

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean:394]: Since it is only used in that lemma, I would put it as an argument there instead of as a variable.

Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: That makes sense!

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: Nirvana, could you resolve all conversations for which you have already implemented the changes? 
Also, I still think that the results specific to norms on the rationals should be moved to a different file, as suggested by Antoine.

Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: Yeah, sorry I didn't know I could do that! (first PR, so much to learn).
I'll make a new file and of course suggestions on different names and stuff are welcome!

Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: There is a working file now, but I suppose I should write some sort of introduction, right?

Comment (fbarroero) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: If I understood @AntoineChambert-Loir right, he was suggesting to include the three lemmas about the rationals directly in the Ostrowski file. I guess it should be called "Ostrowski", but where should it go?  In Analysis/Normed/Ring/Seminorm? Somewhere in NumberTheory?

Comment (riccardobrasca) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: I would say in the number theory folder.

Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: Moved!

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (mariainesdff) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: Maybe the could be moved to the `Rat.MulRingNorm` namespace. So this would become `Rat.MulRingNorm.eq_on_Int_iff_eq_on_Nat`, and similarly for the other lemmas.

Comment (NirvanaC93) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: Uhmm sorry for the silly question, but how do I do it? Do I move the theorems to that file or is there some command to do it in this file? Sorry again 😅

Comment (fbarroero) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: I think it works like a section: you open it with `namespace Rat.MulRingNorm` and close it with `end Rat.MulRingNorm`.

Comment (riccardobrasca) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: I am fixing it.

Comment (riccardobrasca) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: (this is very easy, since it is in a file you don't care about)

Comment (riccardobrasca) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: Done (the bot is not super fast to remove the label, but it will do it). Don't forget to get a new cache after pulling.

Comment (mariainesdff) [reply Mathlib/Analysis/Normed/Ring/Seminorm.lean]: I think `NormRat` should not be part of these three lemma names (since they are already in the `Rat.MulRingNorm` space). So the names should be `eq_on_Int_iff_eq_on_Nat`, `eq_iff_eq_on_Nat` and `equiv_iff_equiv_on_Nat`.
Up to this change, the PR looks good to me, but I think it should be approved by someone not involved with the project.

File: Mathlib/Analysis/Normed/Ring/Seminorm.lean
Comment (faenuccio) [inline Mathlib/Analysis/Normed/Ring/Seminorm.lean:399]:
```suggestion
  obtain ⟨n, rfl | rfl⟩ := eq_nat_or_neg x <;>
  simp only [natAbs_neg, natAbs_ofNat, cast_neg, cast_natCast, map_neg_eq_map]
```

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:19]: Although I appreciate Conrad's works a lot, is it possible to add a more permanent reference (to a book, say, with a proper DOI)?

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:17]: Since this is still ToDo, it would probably be better to add that this is the goal but things remain to be done.

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:27]:
```suggestion
```

Comment (faenuccio) [reply Mathlib/NumberTheory/Ostrowski.lean:27]: Does not seem to be needed...

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:35]: I am not sure that this names adheres to the convention (see https://leanprover-community.github.io/contribute/naming.html). Why is `NormRat` upper case? I cannot find any `structure` in `mathlib` with this name, am I wrong?

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:37]:
```suggestion
  refine ⟨fun h z ↦ ?_, fun a n ↦ a n⟩
```

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:38]:
```suggestion
  obtain ⟨n , rfl | rfl⟩ := eq_nat_or_neg z
```

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:41]:
```suggestion
  · simp only [Int.cast_neg, Int.cast_natCast, map_neg_eq_map, h n]
```

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:33]: The lemma below does not really say this: it says that if two norms coincide on `ℕ` then they coincide on `ℤ` and conversely.

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:47]:
```suggestion
  refine ⟨fun h ↦ ?_, fun h n ↦ congrFun (congrArg DFunLike.coe h) ↑n⟩
```

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:45]: The name of the lemma is somewhat opposite to the statement (which says "equivalent on Nat iff equal"). The same comment applies to the lemma above.

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:58]:
```suggestion
    refine ⟨fun ⟨c, hc, h⟩ ↦ ⟨c, ⟨hc, ?_⟩⟩, fun ⟨c, hc, h⟩ ↦ ⟨c, ⟨hc, fun n ↦ by rw [← h]⟩⟩⟩
```

**Review by faenuccio — COMMENTED** [review]
Thanks for all the beautiful work!

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean:19]: What about Cassels' "Local fields"?

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean:35]: I implemented Maria's suggestion to remove "NormRat" for the name at all!

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean:33]: Don't know why I (or whoever wrote it, I don't remember anymore) wrote it like this, probably a distraction given by the similarity with the other comments. :)

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [reply Mathlib/NumberTheory/Ostrowski.lean:19]: Great choice!

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:22]: For the three references, can you please follow the instructions here
https://leanprover-community.github.io/contribute/doc.html#citing-other-works ?

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:39]:
```suggestion
  obtain ⟨n , rfl | rfl⟩ := eq_nat_or_neg z <;>
  simp only [Int.cast_neg, Int.cast_natCast, map_neg_eq_map, h n]
```

**Review by faenuccio — COMMENTED** [review]
I have added a very minor golfing in one proof; apart from this, please modify the references to adhere to `mathlib` style, expecially because this creates links in the documentation website. After that, I think the PR will be ready to go.

**Comment by riccardobrasca** [comment]
Taking into account the last comments by Filippo LGTM, thanks!

bors d+

**Comment by faenuccio** [comment]
> Taking into account the last comments by Filippo LGTM, thanks!
> 
> bors d+

But please do not forget to modify the reference! You can take inspiration also from the other file you modify
Mathlib/Analysis/Normed/Ring/Seminorm.lean
where they are aligned with Mathlib's conventions.

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean]: Do you want to check before I merge?

Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean]: The bibtool thing doesn't work for me, so I'm not sure I did the correct thing.

**Comment by faenuccio** [comment]
I agree it is a bit painful, but it eventually converges to a working file...

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (riccardobrasca) [reply Mathlib/NumberTheory/Ostrowski.lean]: I think you have to put the item in alphabetical order *manually*

Comment (riccardobrasca) [reply Mathlib/NumberTheory/Ostrowski.lean]: Feel free to complain on zulip :P

Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean]: Happy to do things manually as long as I know that's what I'm supposed to do:)

File: Mathlib/NumberTheory/Ostrowski.lean
Comment (faenuccio) [inline Mathlib/NumberTheory/Ostrowski.lean:50]: I think that also this name should be inverted, right? The `equiv` is the RHS in the statement but appears on the left in the name.

Comment (NirvanaC93) [reply Mathlib/NumberTheory/Ostrowski.lean]: Right. Thanks! Do you know if other than implementing all the bot's changes I have anything else to do? Or will it stop complaining eventually?

**Comment by faenuccio** [comment]
Thanks!

maintainer merge

**Pull request closed without merging**
